Source/Dafny/Rewriter.cs

```diff
@@ -970,6 +970,55 @@ public AutoContractsRewriter(ErrorReporter reporter, BuiltIns builtIns)
     }
   }
 
+  public class OpaqueConstRewriter : IRewriter {
+    public OpaqueConstRewriter(ErrorReporter reporter) : base(reporter) {
+    }
+
+    internal override void PreResolve(ModuleDefinition m) {
+      foreach (var d in m.TopLevelDecls) {
+        if (d is TopLevelDeclWithMembers) {
+          ProcessOpaqueClassConsts((TopLevelDeclWithMembers)d);
+        }
+      }
+    }
+
+    private void ProcessOpaqueClassConsts(TopLevelDeclWithMembers c) {
+      Contract.Requires(c != null);
+      List<MemberDecl> newDecls = new List<MemberDecl>();
+      foreach (MemberDecl member in c.Members) {
+        if (member is ConstantField constant) {
+          if (!Attributes.Contains(constant.Attributes, "opaque")) {
+            // Nothing to do
+          } else if (!RefinementToken.IsInherited(constant.tok, c.EnclosingModuleDefinition)) {
+            GenerateRevealLemma(constant, newDecls);
+          }
+        }
+      }
+      c.Members.AddRange(newDecls);
+    }
+
+    private void GenerateRevealLemma(ConstantField c, List<MemberDecl> newDecls) {
+      // That is, given:
+      //   const {:opaque} foo := x
+      // We produce:
+      //   lemma {:axiom} {:auto_generated} reveal_foo()
+      //     ensures foo == x
+
+      Attributes lemma_attrs = new Attributes("auto_generated", new List<Expression>(), null);
+      lemma_attrs = new Attributes("opaque_reveal", new List<Expression>(), lemma_attrs);
+      lemma_attrs = new Attributes("verify", new List<Expression>() { new LiteralExpr(c.tok, false) }, lemma_attrs);
```

Comment:
I suspect something similar is done for opaque functions. Can the code be shared?